File: apps/web/src/server/repositories/data-gateway.repository.ts
import { AttendanceResult, AttendanceWindow, ClassOffering, Enrollment, Lesson } from '@/types/domain';

/**
 * Repository Layer: Communicates with Google Apps Script data gateway
 */
export interface IDataGatewayRepository {
  getClassOffering(id: string): Promise<ClassOffering | null>;
  saveClassOffering(offering: ClassOffering, roster: Enrollment[], lessons: Lesson[]): Promise<boolean>;
  syncAllClasses(classes: unknown[], startDate: string): Promise<{ success: boolean; spreadsheetUrl?: string }>;
  getRoster(classOfferingId: string): Promise<Enrollment[]>;
  getLesson(id: string): Promise<Lesson | null>;
  getActiveWindow(lessonId: string): Promise<AttendanceWindow | null>;
  openAttendanceWindow(lessonId: string): Promise<AttendanceWindow>;
  closeAttendanceWindow(windowId: string): Promise<boolean>;
  getAttendanceResults(lessonId: string): Promise<AttendanceResult[]>;
  saveCheckIn(lessonId: string, studentEmail: string): Promise<boolean>;
  saveManualOverride(lessonId: string, studentEmail: string, status: 'A' | 'P'): Promise<boolean>;
}

export class GoogleAppsScriptRepository implements IDataGatewayRepository {
  private readonly gatewayUrl: string;

  constructor(gatewayUrl?: string) {
    this.gatewayUrl = gatewayUrl || process.env.APPS_SCRIPT_GATEWAY_URL || '';
  }

  private async postToGateway(action: string, payload: Record<string, unknown>) {
    if (!this.gatewayUrl) {
      // Return stub/fallback when gateway URL is not yet configured
      return { success: true, data: null };
    }

    const response = await fetch(this.gatewayUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ action, payload }),
    });

    if (!response.ok) {
      throw new Error(`Data Gateway returned HTTP ${response.status}`);
    }

    return response.json();
  }

  async getClassOffering(id: string): Promise<ClassOffering | null> {
    const res = await this.postToGateway('getClassOffering', { id });
    return res.data;
  }

  async saveClassOffering(offering: ClassOffering, roster: Enrollment[], lessons: Lesson[]): Promise<boolean> {
    const res = await this.postToGateway('saveClassOffering', { offering, roster, lessons });
    return res.success;
  }

  async syncAllClasses(classes: unknown[], startDate: string): Promise<{ success: boolean; spreadsheetUrl?: string }> {
    const res = await this.postToGateway('syncAllClasses', { classes, startDate });
    return {
      success: res.success,
      spreadsheetUrl: res.data?.spreadsheetUrl,
    };
  }

  async getRoster(classOfferingId: string): Promise<Enrollment[]> {
    const res = await this.postToGateway('getRoster', { classOfferingId });
    return res.data || [];
  }

  async getLesson(id: string): Promise<Lesson | null> {
    const res = await this.postToGateway('getLesson', { id });
    return res.data;
  }

  async getActiveWindow(lessonId: string): Promise<AttendanceWindow | null> {
    const res = await this.postToGateway('getActiveWindow', { lessonId });
    return res.data;
  }

  async openAttendanceWindow(lessonId: string): Promise<AttendanceWindow> {
    const res = await this.postToGateway('openAttendanceWindow', { lessonId });
    return res.data;
  }

  async closeAttendanceWindow(windowId: string): Promise<boolean> {
    const res = await this.postToGateway('closeAttendanceWindow', { windowId });
    return res.success;
  }

  async getAttendanceResults(lessonId: string): Promise<AttendanceResult[]> {
    const res = await this.postToGateway('getAttendanceResults', { lessonId });
    return res.data || [];
  }

  async saveCheckIn(lessonId: string, studentEmail: string): Promise<boolean> {
    const res = await this.postToGateway('saveCheckIn', { lessonId, studentEmail });
    return res.success;
  }

  async saveManualOverride(lessonId: string, studentEmail: string, status: 'A' | 'P'): Promise<boolean> {
    const res = await this.postToGateway('saveManualOverride', { lessonId, studentEmail, status });
    return res.success;
  }
}

// Export singleton instance
export const dataGatewayRepository = new GoogleAppsScriptRepository();

File: backend/lib/repositories/data_gateway_repository.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config.dart';

class DataGatewayRepository {
  final String gatewayUrl;

  DataGatewayRepository({String? gatewayUrl})
      : gatewayUrl = gatewayUrl ?? AppConfig.appsScriptGatewayUrl;

  Future<Map<String, dynamic>> _postToGateway(String action, Map<String, dynamic> payload) async {
    if (gatewayUrl.isEmpty) {
      return {'success': true, 'data': null};
    }

    var response = await http.post(
      Uri.parse(gatewayUrl),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'action': action,
        'payload': payload,
      }),
    );

    // Google Apps Script Web Apps respond with 302 Redirect to usercontent
    if (response.statusCode == 302 || response.statusCode == 301 || response.statusCode == 303 || response.statusCode == 307) {
      final redirectUrl = response.headers['location'];
      if (redirectUrl != null && redirectUrl.isNotEmpty) {
        response = await http.get(Uri.parse(redirectUrl));
      }
    }

    if (response.statusCode != 200) {
      throw Exception('Data Gateway returned HTTP ${response.statusCode}: ${response.body}');
    }

    final dynamic decoded = jsonDecode(response.body);
    if (decoded is Map<String, dynamic>) {
      return decoded;
    }
    return {'success': true, 'data': decoded};
  }

  Future<Map<String, dynamic>> syncAllClasses(List<dynamic> classes, String startDate) async {
    final res = await _postToGateway('syncAllClasses', {
      'classes': classes,
      'startDate': startDate,
    });
    final success = res['success'] == true && res['data'] != false && res['data'] != null;
    print('[DataGateway] syncAllClasses: success=$success');
    return {
      'success': success,
      'spreadsheetUrl': res['data']?['spreadsheetUrl'],
      'message': 'Đã đồng bộ toàn bộ các lớp học lên Google Sheet thành công.',
    };
  }

  Future<bool> saveCheckIn(String lessonId, String studentEmail) async {
    final res = await _postToGateway('saveCheckIn', {
      'lessonId': lessonId,
      'studentEmail': studentEmail,
    });
    final ok = res['success'] == true && res['data'] != false && res['data'] != null;
    print('[DataGateway] saveCheckIn ($lessonId, $studentEmail) -> ${ok ? "SUCCESS" : "FAILED"} (res: $res)');
    return ok;
  }

  Future<bool> saveManualOverride(String lessonId, String studentEmail, String status) async {
    final res = await _postToGateway('saveManualOverride', {
      'lessonId': lessonId,
      'studentEmail': studentEmail,
      'status': status,
    });
    final ok = res['success'] == true && res['data'] != false && res['data'] != null;
    print('[DataGateway] saveManualOverride ($lessonId, $studentEmail, $status) -> ${ok ? "SUCCESS" : "FAILED"} (res: $res)');
    return ok;
  }

  Future<List<Map<String, dynamic>>> getAttendanceResults(String lessonId) async {
    final res = await _postToGateway('getAttendanceResults', {
      'lessonId': lessonId,
    });
    final dynamic data = res['data'];
    if (data is List) {
      return data.cast<Map<String, dynamic>>();
    }
    return [];
  }

  Future<Map<String, dynamic>?> openAttendanceWindow(String lessonId) async {
    final res = await _postToGateway('openAttendanceWindow', {
      'lessonId': lessonId,
    });
    return res['data'] as Map<String, dynamic>?;
  }

  Future<bool> closeAttendanceWindow(String windowId) async {
    final res = await _postToGateway('closeAttendanceWindow', {
      'windowId': windowId,
    });
    return res['success'] == true;
  }

  Future<Map<String, dynamic>?> getActiveWindow(String lessonId) async {
    final res = await _postToGateway('getActiveWindow', {
      'lessonId': lessonId,
    });
    return res['data'] as Map<String, dynamic>?;
  }
}
